### PR TITLE
Feature/exit code

### PR DESCRIPTION
**Pull request by garyluu**
This fixes exit code by throwing runtime exceptions when the process exit code is not 0.
Now printing every command s3cmd is executing to assist in further debugging.
Small printing fix to reduce output lines.

### CODE REVIEW

**Comment by denis-yuen** [comment]
FYI, noticed a possible bug in the plugin handling code for uploads. Let's rework this to throw exceptions on non-recoverable errors (like a bucket not being present or something using ctrl-c) but return false for recoverable errors (like if an upload aborts due to low speed)

**Comment by denis-yuen** [comment]
Test

**Comment by garyluu** [comment]
Actually, my plugin actually creates the bucket if it doesn't exist.  Which ones are recoverable errors aside from upload aborts due to low speed (which I'm not sure how to check)

**Comment by denis-yuen** [comment]
Maybe get a list of errors from S3cmd and do a first pass here.

**Comment by garyluu** [comment]
List of exit codes from the s3cmd client itself:

EX_OK               = 0
EX_GENERAL          = 1
EX_PARTIAL          = 2    # some parts of the command succeeded, while others failed
EX_SERVERMOVED      = 10   # 301: Moved permanantly & 307: Moved temp
EX_SERVERERROR      = 11   # 400, 405, 411, 416, 417, 501: Bad request, 504: Gateway Time-out
EX_NOTFOUND         = 12   # 404: Not found
EX_CONFLICT         = 13   # 409: Conflict (ex: bucket error)
EX_PRECONDITION     = 14   # 412: Precondition failed
EX_SERVICE          = 15   # 503: Service not available or slow down
EX_USAGE            = 64   # The command was used incorrectly (e.g. bad command line syntax)
EX_DATAERR          = 65   # Failed file transfer, upload or download
EX_SOFTWARE         = 70   # internal software error (e.g. S3 error of unknown specificity)
EX_OSERR            = 71   # system error (e.g. out of memory)
EX_OSFILE           = 72   # OS error (e.g. invalid Python version)
EX_IOERR            = 74   # An error occurred while doing I/O on some file.
EX_TEMPFAIL         = 75   # temporary failure (S3DownloadError or similar, retry later)
EX_ACCESSDENIED     = 77   # Insufficient permissions to perform the operation on S3
EX_CONFIG           = 78   # Configuration file error
_EX_SIGNAL          = 128
_EX_SIGINT          = 2
EX_BREAK = _EX_SIGNAL + _EX_SIGINT # Control-C (KeyboardInterrupt raised)

**Comment by denis-yuen** [comment]
And as your first pass, which ones would you consider recoverable and which ones not?

**Comment by garyluu** [comment]
Probably just 1 and 75, everything else requires something external to be changed (config files, credentials, file locations)

File: src/main/java/io/dockstore/provision/S3CmdPlugin.java [hunk 1]
```diff
@@ -99,10 +99,20 @@ public boolean downloadFrom(String sourcePath, Path destination) {
             sourcePath = sourcePath.replaceFirst("s3cmd", "s3");
             String command = client + " -c " + configLocation + " get " + sourcePath + " " + destination + " --force";
             int exitCode = executeConsoleCommand(command, true);
-            if (exitCode != 0) {
+            switch (exitCode) {
```
Comment (denis-yuen) [inline src/main/java/io/dockstore/provision/S3CmdPlugin.java]: Please throw in a few comments and link to the source code where you found descriptions of the exit codes for s3cmd

**Pull request merged**